### PR TITLE
Pre commit hook

### PR DESCRIPTION
**Pull request by Noddy20**
## Issue
- NA : No issue present, contribution to improve CI pipeline experience

## Overview
- Added pre commit git hook to run the detekt and lint check. Commit will only succeed if the pre commit checks passes.
- This helps preventing CI pipeline runs that might fail due to minor formatting or lint related issue
- This also helps first time contributors to fix all the formatting related issues before committing and pushing any change

## Links
- NA

## Screenshot (Optional if screenshot test is present or unrelated to UI)
Before | After
:--: | :--:
NA

## Movie (Optional)
Before | After
:--: | :--:
NA

### CODE REVIEW

**Comment by takahirom** [comment]
Thank you. I didn't know that you could insert a hook like this with Gradle. It seems like it might also handle Gradle's cache, which is nice. This method was even discussed among the staff, but while formatting with hooks and such is certainly one way to do it, I find it a bit awkward to be kept waiting with every small commit. There might be cases where you want to format everything at once, so I think I won't merge this time. I'm really sorry, but I think it will be a good reference, so thank you.

File: git-hooks.gradle.kts
Comment (jmatsu) [inline git-hooks.gradle.kts:24]: You can change the git hooks directory from `.git/hooks` to any directory you want. It's better to use the custom hook directory rather than copying hook scripts frequently.

Please add these lines to /.gitignore.

```
# Ignore files under /hooks except pre-defined hooks
/hooks/*
!/hooks/pre-commit
```

and then, create a task in the */build.gradle.kts*.

```suggestion
tasks.register("installGitHooks", Exec::class.java) {
    description = "Set up local git config to use /hooks for git-hooks"
    group = "git hooks"
    workingDir = layout.projectDirectory.getAsFile()
    commandLine = listOf("git")
    args("config", "--local", "core.hooksPath", "hooks")
}
```

Lastly, please make reference to `./gradlew installGitHooks` in [README.md#Requirements](https://github.com/DroidKaigi/conference-app-2023#requirements).

File: hooks/pre-commit
Comment (jmatsu) [inline hooks/pre-commit:2]: `exec 1>&2` is a better practice that redirects stdout to stderr in hooks because some of users may wanna use stdout of `git commit`.

```suggestion
set -e

exec 1>&2
```

File: hooks/pre-commit
Comment (jmatsu) [inline hooks/pre-commit:15]: `-e` option means this script(shell) will abort right after any non-zero code is returned so `$?` without if-statement won't work. 

```suggestion
if ./gradlew lintDevDebug then
    echo "No issues were found in lintDevDebug"
else
    echo "Android Lint exited with $?. Need to fix issues before commit."
    exit 1
fi
```

File: hooks/pre-commit
Comment (jmatsu) [inline hooks/pre-commit:27]:
```suggestion
if ./gradlew detekt then
    echo "No issues were found, committing changes."
else
    echo "Detekt exited with $?. Need to fix issues before commit."
    exit 1
fi
```

File: build.gradle.kts
Comment (jmatsu) [inline build.gradle.kts:13]: Please define the install task directly.

```suggestion
```

**Review by jmatsu — CHANGES_REQUESTED** [review]
Although we still continue to discuss this topic as described in https://github.com/DroidKaigi/conference-app-2023/pull/916#issuecomment-1689953649, I personally love Git hooks solution so let me submit a couple of reviews in advance.

File: build.gradle.kts
Comment (Noddy20) [reply build.gradle.kts:13]: Any reason for defining it directly?
I kept a separate file for clean code (Separation of Concerns) and better visibility.

Comment (takahirom) [reply build.gradle.kts:13]: I think you can refer to this.
https://developer.squareup.com/blog/herding-elephants/#model-your-build

**Comment by takahirom** [comment]
We might use these scripts on other projects but for now, we don't use this. Let me close. Thank you for your suggestion and work.

**Pull request closed without merging**
